### PR TITLE
azure - add azure resource graph as a query provider

### PR DESCRIPTION
**Pull request by aluong**
Disclaimer: The resource graph does not return exactly the same response as ARM.

- This is an initial plumbing for using the Azure Resource Graph as a provider

### CODE REVIEW

File: tools/c7n_azure/tests_azure/test_resource_graph_source.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                 'value_type': 'normalize',
+                 'value': 'cctstorage*'}]
+        })
+
+        resources_arm = p1.run()[0]
+
+        p2 = self.load_policy({
+            'name': 'test-azure-storage-arm-source',
+            'resource': 'azure.storage',
+            'source': 'resource-graph',
```
Comment (stefangordon) [inline tools/c7n_azure/tests_azure/test_resource_graph_source.py]: I am curious if there is precedent in this schema for us to add additional fields about the source?  Like some sort of filter query string...

File: tools/c7n_azure/tests_azure/test_resource_graph_source.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            'name': 'test-azure-vm-arm-source',
+            'resource': 'azure.vm',
+            'source': 'resource-graph',
+            'filters': [
+                {'type': 'value',
+                 'key': 'name',
+                 'op': 'glob',
+                 'value_type': 'normalize',
+                 'value': 'cctestvm*'}]
+        }, validate=True)
```
Comment (stefangordon) [inline tools/c7n_azure/tests_azure/test_resource_graph_source.py]: note we are trying to go back and ensure we only do validate true once, usually in a dedicated validate test method because of how slow it is (adds 2 seconds to each test).

**Comment by stefangordon** [comment]
Cool! Excited to start experimenting with this.

File: tools/c7n_azure/tests_azure/test_resource_graph_source.py (hunk 1 above)
Comment (kapilt) [reply tools/c7n_azure/tests_azure/test_resource_graph_source.py]: there's a query facility that enables it, we use it to pass sql select statements with source: config in aws. I think I demo'd with this Andy at pycon hackathon when the graph provider was being worked on.

**Comment by aluong** [comment]
Test failure is due to how Py27 unicode and bytestrings. Resource graph is getting deserialized into unicode strings oppose to bytestrings. This problem goes away in Py3+.

File: tools/c7n_azure/tests_azure/test_resource_graph_source.py (hunk 1 above)
Comment (aluong) [reply tools/c7n_azure/tests_azure/test_resource_graph_source.py]: Totally can do this. Let's add at as an enhancement after we get this in and play with it more.

File: tools/c7n_azure/c7n_azure/query.py [hunk 3]
```diff
@@ -89,6 +90,34 @@ def augment(self, resources):
         return resources
 
 
+@sources.register('resource-graph')
+class ResourceGraphSource(object):
+
+    def __init__(self, manager):
+        self.manager = manager
+
+    def validate(self):
```
Comment (kapilt) [inline tools/c7n_azure/c7n_azure/query.py]: do we need a minimal validation here that we have the appropriate resource type for graph client usage for the policy's resource? are all custodian resources supported by resource graph client?

Comment (aluong) [reply tools/c7n_azure/c7n_azure/query.py]: Great point. I added validation to now allow processing of resources that we cannot support. An example would be the data plane child resources such as `azure.keyvault-certificate`

File: tools/c7n_azure/tests_azure/test_resource_graph_source.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        for prop in res1:
+            if prop not in ignore_properties and \
+                    (prop not in res2 or not resource_cmp(res1[prop], res2[prop])):
+                return False
+
+    elif type(res1) == list:
+        if len(res1) != len(res2):
+            return False
+
+        for item1, item2 in zip(res1, res2):
```
[the remark this answers is not in the document]
Comment (aluong) [reply tools/c7n_azure/tests_azure/test_resource_graph_source.py]: Agreed, this gets a little tricky because sometimes the element of the array is another dictionary and it's not possible to know the key ahead of time.

File: tools/c7n_azure/c7n_azure/query.py [hunk 5]
```diff
@@ -79,6 +81,9 @@ def __init__(self, manager):
         self.manager = manager
         self.query = self.resource_query_factory(self.manager.session_factory)
 
+    def validate(self):
```
Comment (erwelch) [inline tools/c7n_azure/c7n_azure/query.py]: Why is this validate method needed?

Comment (aluong) [reply tools/c7n_azure/c7n_azure/query.py]: https://github.com/cloud-custodian/cloud-custodian/pull/4924#discussion_r333024918
tldr; not all resources can be queried by resource-graph

Comment (kapilt) [reply tools/c7n_azure/c7n_azure/query.py]: but shouldn't it be actually validating that then?
